File: 2023/02.js
const fs = require("fs");
const input = fs
  .readFileSync("02.txt", "utf8")
  .split("\n")
  .filter((l) => l.length);

const part1 = () => {
  const cubes = {
    red: 12,
    green: 13,
    blue: 14,
  };

  return input
    .map((line) => {
      const game = parseGame(line);
      let valid = game.num;

      Object.keys(cubes).forEach((key) => {
        game.hands.forEach((hand) => {
          if (hand[key] > cubes[key]) {
            valid = 0;
          }
        });
      });

      return valid;
    })
    .reduce((a, b) => a + b, 0);
};

const part2 = () => {
  return input
    .map((line) => {
      const game = parseGame(line);
      const leastCubes = {
        red: 0,
        blue: 0,
        green: 0,
      };

      game.hands.forEach((hand) => {
        Object.keys(hand).forEach((key) => {
          if (hand[key] > leastCubes[key]) {
            leastCubes[key] = hand[key];
          }
        });
      });

      return leastCubes.red * leastCubes.green * leastCubes.blue;
    })
    .reduce((a, b) => a + b, 0);
};

const parseGame = (line) => {
  const game = line.split(":")[0].split(" ")[1].trim();
  const hands = line.split(":")[1].trim().split(";");

  return {
    num: Number(game),
    hands: hands.map((h) => parseHand(h)),
  };
};

const parseHand = (hand) => {
  const cubes = hand.split(",").map((h) => h.trim());

  return {
    green: getCubeUsages(cubes, "green"),
    red: getCubeUsages(cubes, "red"),
    blue: getCubeUsages(cubes, "blue"),
  };
};

const getCubeUsages = (cubes, color) => {
  return cubes
    .filter((p) => p.endsWith(color))
    .map((g) => Number(g.match(/\d+/g)[0]))
    .reduce((a, b) => a + b, 0);
};

console.time("part1");
console.log(part1());
console.timeEnd("part1");

console.time("part2");
console.log(part2());
console.timeEnd("part2");

File: 2023/02.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Game struct {
	num   int
	hands []Hand
}

type Hand struct {
	red   int
	blue  int
	green int
}

const (
	redLimit   = 12
	greenLimit = 13
	blueLimit  = 14
)

func main() {
	part1()
}

func part1() {
	lines := readInput()
	var games []Game

	for _, line := range lines {
		games = append(games, parseLine(line))
	}

	total := calculatePart1Total(games)

	fmt.Println(total)
}

func calculatePart1Total(games []Game) int {
	total := 0

	for _, game := range games {
		if isValidGame(game) {
			total += game.num
		}
	}

	return total
}

func isValidGame(game Game) bool {
	for _, hand := range game.hands {
		if hand.red > redLimit || hand.green > greenLimit || hand.blue > blueLimit {
			return false
		}
	}

	return true
}

func parseLine(line string) Game {
	lineParts := strings.Split(line, ":")
	gameParts := strings.Split(lineParts[0], " ")
	gameHands := strings.Split(lineParts[1], ";")

	game := Game{}
	game.num, _ = strconv.Atoi(gameParts[1])

	for _, hand := range gameHands {
		cubes := strings.Split(hand, ",")

		newHand := Hand{}
		for _, cube := range cubes {
			cubeParts := strings.Fields(cube)

			switch cubeParts[1] {
			case "red":
				newHand.red, _ = strconv.Atoi(cubeParts[0])
			case "blue":
				newHand.blue, _ = strconv.Atoi(cubeParts[0])
			case "green":
				newHand.green, _ = strconv.Atoi(cubeParts[0])
			}
		}

		game.hands = append(game.hands, newHand)
	}

	return game
}

func readInput() []string {
	input, err := os.Open("02.txt")
	if err != nil {
		panic(err)
	}
	defer input.Close()

	fileScanner := bufio.NewScanner(input)
	fileScanner.Split(bufio.ScanLines)
	var fileLines []string

	for fileScanner.Scan() {
		line := fileScanner.Text()

		if len(line) > 0 {
			fileLines = append(fileLines, line)
		}
	}

	return fileLines
}
